### PR TITLE
Query Action implementation & related refactoring

### LINKED ISSUE #20726: Implementation of do-action for streams
Spec to track - https://github.com/ballerina-platform/ballerina-spec/issues/403

### PR DESCRIPTION
**Pull request by mohanvive**
Fixes #20726

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+        //      x[x.length()] = {
+        //         firstName: person.firstName,
+        //         lastName: person.lastName
+        //      };
+
+        if (selectClause.expression.type == null) {
+            selectClause.expression.type = fromClause.varType;
+        }
+
+        BLangInvocation lengthInvocation = createLengthInvocation(selectClause.pos, outputArrayVariable.symbol);
```
Comment (gimantha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java]: Shall we replace this with a simple array:push function?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java [hunk 2]
```diff
@@ -1160,6 +1167,27 @@ public void visit(BLangForeach foreach) {
         analyzeExpr(foreach.collection);
     }
 
+    @Override
+    public void visit(BLangQueryAction queryAction) {
+        this.loopWithintransactionCheckStack.push(true);
+        boolean statementReturns = this.statementReturns;
+        this.checkStatementExecutionValidity(queryAction);
+        this.loopCount++;
+        queryAction.doClause.accept(this);
```
Comment (gimantha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java]: without directly calling accept() shall we call analyzeExpr()

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java [hunk 3]
```diff
@@ -449,6 +456,19 @@ public void visit(BLangForeach foreach) {
         analyzeNode(foreach.body, env);
     }
 
+    @Override
+    public void visit(BLangQueryAction queryAction) {
+        for (FromClauseNode fromClauseNode : queryAction.fromClauseList) {
+            ((BLangFromClause) fromClauseNode).accept(this);
```
Comment (gimantha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java]: Use analyzeNode instead

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java (hunk 1 above)
Comment (mohanvive) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java]: Facing an issue for primitive arrays @ runtime when using push method. Will attend this separately.

**Pull request merged**